consensus/service/src/client_api_service.rs

```diff
@@ -26,135 +25,468 @@ use std::sync::Arc;
+        let mut ledger = MockLedger::new();
+        // The service should request num_blocks.
+        ledger
+            .expect_num_blocks()
+            .times(1)
+            .return_const(Ok(num_blocks));
+
+        // The service should check if each key image is already in the ledger.
+        ledger
+            .expect_contains_key_image()
+            .times(2)
+            .return_const(Ok(false));
+
+        let mut tx_manager = MockTxManager::new();
+        tx_manager.expect_insert().times(1).return_const(Err(
+            TxManagerError::TransactionValidation(TransactionValidationError::InvalidRangeProof),
+        ));
+
+        let is_serving_fn = Arc::new(|| -> bool { true });
+
+        let instance = ClientApiService::new(
+            Arc::new(consensus_enclave),
+            scp_client_value_sender,
+            Arc::new(ledger),
+            Arc::new(tx_manager),
+            is_serving_fn,
+            logger,
+        );
+
+        // gRPC client and server.
+        let (client, _server) = get_client_server(instance);
+
+        let message = Message::default();
+        match client.client_tx_propose(&message) {
+            Ok(propose_tx_response) => {
+                assert_eq!(
+                    propose_tx_response.get_result(),
+                    ProposeTxResult::InvalidRangeProof
+                );
+                assert_eq!(propose_tx_response.get_num_blocks(), num_blocks);
+            }
+            Err(e) => panic!("Unexpected error: {:?}", e),
+        }
+    }
+
+    #[test_with_logger]
+    #[serial(counters)]
+    // Should return RpcStatus Unavailable if the node is not serving.
+    fn test_client_tx_propose_tx_is_not_serving(logger: Logger) {
+        let mut enclave = MockConsensusEnclave::new();
+        enclave
+            .expect_client_discard_message()
+            .times(1)
+            .return_const(Ok(()));
+
+        let is_serving_fn = Arc::new(|| -> bool { false }); // Not serving
+
+        let scp_client_value_sender = Arc::new(
+            |_tx_hash: TxHash, _node_id: Option<&NodeID>, _responder_id: Option<&ResponderId>| {},
+        );
+
+        let instance = ClientApiService::new(
+            Arc::new(enclave),
+            scp_client_value_sender,
+            Arc::new(MockLedger::new()),
+            Arc::new(MockTxManager::new()),
+            is_serving_fn,
+            logger,
+        );
+
+        // gRPC client and server.
+        let (client, _server) = get_client_server(instance);
+
+        let message = Message::default();
+        match client.client_tx_propose(&message) {
+            Ok(propose_tx_response) => {
+                panic!("Unexpected response {:?}", propose_tx_response);
+            }
+            Err(_) => {
+                // Should be RpcFailure(RpcStatus { status: 14-UNAVAILABLE, details: Some("Temporarily not serving requests")
+            }
+        }
+    }
+
+    #[test_with_logger]
+    #[serial(counters)]
```

Comment:
Unfortunately, these tests share global state via `counters`, so they need to run serially. Maybe we can change that later, but I'm not interested in dealing with it now.